# Task 1

Write a Python function called sqrt2 that calculates and
prints to the screen the square root of 2 to 100 decimal places. Your code should
not depend on any module from the standard library1 or otherwise. You should
research the task first and include references and a description of your algorithm.

# Code

In [2]:
x = 2 * 10 ** 200

#storing the original x value in y
y = x

#storing a value for 10**100
power = 10**100

def sqrt2(x,y):
    originalDer = abs(x - y**2)
    minusDer = abs(x - (y-1)**2)
    plusDer= abs(x - (y+1)**2)
    
    minimised = originalDer <= minusDer and originalDer <= plusDer
    below_min =  plusDer < minusDer
    
    return minimised, below_min

while True:
    oldy = y
    y = (y + x // y) // 2
    minimised, below_min = sqrt2(x, y)
    if minimised:
        break

    if y == oldy:
        if below_min:
            y += 1
        else:
            y -= 1
        minimised, _ = sqrt(x, y)
        if minimised:
            break

print(f'{y // power}.{y % power:100d}')
#code by alani

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


# Previous Iterations

In [3]:
#function to run newtons method to calculate the square root of 2
def sqRoot(number,iterations):
    #Store input of n in x
    x = float(number)
    #variable to count the iterations
    count = 0 
    #for loop to run the number of iterations
    for count in range(iterations):
        number = 0.5 * (number + x / number)
    return number

print("{0:.100f}".format(sqRoot(2,10)))

1.4142135623730949234300169337075203657150268554687500000000000000000000000000000000000000000000000000


In [8]:
#Using the Babylonian Method
def baby_method(num):
    x = num
    y = 1 
    maxError = 0.0000001
    while x - y > maxError:
        x = (x + y)/2;
        y = num/x;
    return x
print("The square root of 2 is : {0:.100f}".format(baby_method(2)))

The square root of 2 is : 1.4142135623746898698271934335934929549694061279296875000000000000000000000000000000000000000000000000


# Analysis

This is not my final algorithm. This is right now the best answer I have found. I have tried the babylonian and newtons method for computing algorithms but to no avail. I will come back to this soon. The algorithm was found on stack overflow from a use named alani. I modified a few pieces but no enough to call it my own. The code begins by setting the value of x to 2, then times it by 10 before exponentiate that answer by 200.This is to override the floating number precision limit of 53 decimal places. The function generate_derivatives is used to calculate true or false values for the wile loop.

The two attempts shown are defined functions I coded to run find the square root of 2. Due to Float having a limit of 53 decimal places the output is not correct. These are previous attempts I wrote of the assignment that led to the best outcome.I am not finished or happy with this yet.

# References

Powers function in python:https://docs.python.org/3/tutorial/introduction.html <br />
Square root:https://www.tutorialspoint.com/How-to-perform-square-root-without-using-math-module-in-Python <br />
formating python:https://www.geeksforgeeks.org/precision-handling-python/ <br />
code: https://stackoverflow.com/questions/64278117/is-there-a-way-to-create-more-decimal-points-on-python-withoutimporting-a-libra <br />
Undestanding Newtons method: https://www.math.ubc.ca/~pwalls/math-python/roots-optimization/newton/ <br />
    https://www.youtube.com/watch?v=2158QbsunA8&ab_channel=LorenzoSadun <br />
Understanding Babylonian Method: https://blogs.sas.com/content/iml/2016/05/16/babylonian-square-roots.html#:~:text=The%20Babylonian%20square%2Droot%20algorithm,Newton%20invented%20his%20general%20procedure <br />

# Task 2

The Chi-squared test for independence is a statistical hypothesis test like a t-test. It is used to analyse whether two categorical variables are independent. The Wikipedia article gives the table below as an example [4], stating the Chi squared value based on it is approximately 24.6. Use scipy.stats to verify this value and calculate the associated p value. You should include a short note with references justifying your analysis in a markdown cell.

# Table

In [9]:
headings = ["A","B","C","D"]
data =[[ 90, 60, 104, 95],[ 30, 50, 51, 20],[30, 40, 45, 35]]
format_row = "{:>10}" * (len(headings) + 1)
print(format_row.format("", *headings))
for team, row in zip(headings, data):
    print(format_row.format(team, *row)) 

                   A         B         C         D
         A        90        60       104        95
         B        30        50        51        20
         C        30        40        45        35


# Code

In [ ]:
import scipy.stats as stats
  
# data stored in a 2-D array from the table provided
data = [[ 90, 60, 104, 95],[ 30, 50, 51, 20],[30, 40, 45, 35]]
#set a significant value
sigValue = 24.6
stat, p, dof, expected = stats.chi2_contingency(data) 

print("p value is " + str(p)) 

print("Expected")
print(expected)

print("Dof = ", dof)
print("Stat = ", stat)

# Analysis

The result received in this test is 24.57. This is lower than the significant value of 24.6. The difference is 0.03. A null hypothesis can now not be accepted because we can see that there is a relationship between the 2 categories. The value we get for p also reaffirms this conclusion.

# References

Undestanding Chi-Square test: https://stattrek.com/chi-square-test/independence.aspx
https://www.investopedia.com/terms/c/chi-square-statistic.asp#:~:text=A%20chi%2Dsquare%20(%CF%872,compares%20to%20actual%20observed%20data.&text=The%20chi%2Dsquare%20statistic%20compares,of%20variables%20in%20the%20relationship)
Understanding desgrees of freedom:https://statisticsbyjim.com/hypothesis-testing/degrees-freedom-statistics/
Scipy docs:https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html

# Task 3

The standard deviation of an array of numbers x is calculated using numpy as np.sqrt(np.sum((x - np.mean(x))** 2)/len(x)).
However, Microsoft Excel has two different versions of the standard deviation calculation, STDEV.P and STDEV.S .
The STDEV.P function performs the above calculation but in the STDEV.S calculation the division is by len(x)-1 rather
than len(x) . Research these Excel functions, writing a note in a Markdown cell about the difference between them. Then use numpy to perform a simulation demonstrating that the STDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample. 
Note that part of this taskis to figure out the terminology in the previous sentence.

# Research

Understanding Stdev in excel:https://www.youtube.com/watch?v=NdIVwuwfMIg&ab_channel=DesigneLearningTutorials <br />
http://mathbitsnotebook.com/Algebra1/StatisticsData/STPopSample.html <br />
Stdev.p is used to find the standard deviation of all the data acquired. The p stands for population. In excel, you run the formula while selecting all data cells in the spreadsheet to find the result.
Stdev.s is used to calculate the standard deviation of a sample of the data required. In excel you choose a select number of data cells and run the formula to find the standard deviation of those selected cells.

# Code

# Standard Deviation Population Test

In [21]:
import numpy as np
#data is random results of a test in % format - 25 in total
data = ([75,22,82,48,51,92,72,64,38,57,62,79,88,28,53,83,75,66,43,52,78,91,85,34,58])
print("standard deviation population result")
#formula for sample - np.sqrt(np.sum((x - np.mean(x))**2)/len(x))
np.sqrt(np.sum((data - np.mean(data))**2)/len(data))

standard deviation population result


19.777724843874232

# Standard Deviation Sample Test

In [5]:
import numpy as np
#data is random results of a test in % format - 25 in total
data = ([75,22,82,48,51,92,72,64,38,57,62,79,88,28,53,83,75,66,43,52,78,91,85,34,58])
print("standard deviation sample result")
np.sqrt(np.sum((data - np.mean(data))**2)/len(data)-1)

standard deviation sample result


19.75242769889312

# Analysis

The results of both tests gives the user a basis of where the standard deviation of the tests is based. Stdev.p = 19.78(to 2 decimal places) while stdev.s = 17.75(to two decimal places). The difference to two decimal places is 0.03. This gives the user an understanding that the data is mainly between 20% of the mean.
Using Stdev.s gives the user a better understanding of what the standard deviaton is because it will offer a greater fluctuation of results due it taking only select results that are randomly choosen from the data provided.When the data provided gets larger, the population result can change rapidly because alot of people could score around the same mark.This will skew the data and give out a result that may not accurately reflect the data provided to give the best outcome. Due to this using the sample standard deviation formula allows the data to be randomly choosen and give a better outcome because the data provided will be of a greater variance. 
With the data provided in the above code size was not 25 but 2500 and 5-% of the results are between 45-55, this will cause the standard dev population to be a much lower number than the number calculated in this file.

# Task 4

Use scikit-learn to apply k-means clustering to Fisher’s famous Iris data set. You will easily obtain a copy of the data set online. Explain in a Markdown cell how your code works and how accurate it might be, and then explain how your model could be used to make predictions of species of iris.

# Code

In [3]:
from sklearn.cluster import KMeans
import numpy as np
#Fisher's Iris Dataset
iris = datasets.load_iris()

X = np.array([[1, 2], [1, 4], [1, 0],
              [10, 2], [10, 4], [10, 0]])
kmeans = KMeans(n_clusters=2, random_state=0).fit(X)
kmeans.labels_

kmeans.predict([[0, 0], [12, 3]])

kmeans.cluster_centers_

array([[10.,  2.],
       [ 1.,  2.]])